### ISSUE TITLE
RMC is slower than PMC w.r.t. mesh update

### ISSUE THREAD

**Issue by codeonwort** [Koderz/RuntimeMeshComponent#88]
Hi.

README says RMC is far faster than PMC(ProceduralMeshComponent) to update mesh data, but RMC is actually super slow when there's many sections in the same component.

Let's say a RMC represents a grid of cells and each cell is represented by a section. When I only update one section at a time, RMC is too slow to put on practical use. PMC doesn't have this problem even though there's many sections.

I don't know how exactly RMC works, but looking into the code, when I update just one section it seems some works are done for all sections such as rendering and collision update. (entire bound update is also done, but it's done in PMC in the same way, so it's not an important factor I guess.)

**Comment by Koderz** [comment]
Hey @codeonwort 

What version of the RMC/Engine are you using?  Also how are you determining it to be slower?  I'll admit it's been a while since I profiled it, but there's no reason it should have gotten any slower than the PMC to do the same thing.   The comment I'd make though is there's 2+ threads involved in both components, so you have to watch multiple timings to see what's actually going on.

Both components will have to update collision the same way usually, that's an engine limitation, but things like async cooking setting would have to be the same between the two to get any real accurate view.   The PMC uses a faster path to update collision for a section, but it also assumes the size of a section never changes, which the RMC doesn't.  I might add the dynamic choice for it to use whichever makes more sense but I was more focused on getting the core functional, then try to optimize it. Unfortunately right now though I have to spend time redoing some of the core due to 4.19 changes, so I'll have to do that before I can really start optimizing parts of it.

Both components will recreate the render proxy on creation/removal of a section, but the RMC doesn't copy all the sections buffer data to the GPU again each time, so that's definitely faster in the RMC. The RMC just recreates the proxy on section updates since it doesn't have to sync all the buffer data so it's a smaller improvement to do async updates like the PMC does.  I'll still likely add it, just wasn't a high priority during the rewrite.

Due to some coming 4.19  changes, much of that is actually only going to get faster in the RMC, at the expense of some of its flexibility and a change in mesh creation/update API if you want the fastest speed.

If you want to share some of the timings your seeing as well as RMC/Engine version I'll gladly look into what might be causing it and if there's something I can do within the RMC to improve it.

Also, I don't usually recommend having many sections in one component. There's a couple reasons for that. First, any time you create/remove a section it will have to perform a full collision cook of the entire component (same as PMC)  which can be extremely slow. The second is rendering performance... While I can't give a hard number, there's the problem that every section in a component is combined into a single entry in the renderer so they all get culled/occluded together, so if one's visible they're all visible.   You have the same number of draw calls if it's 10 sections in one component vs 1 section in each of 10 components.

**Comment by codeonwort** [comment]
I tested it on UE4.18.3 and current master branch of RMC. By slow I mean when I update a section among, say 16x16 sections, the framerate goes down to 15 fps.

What I did was that, I tried to implement some kind of dynamic landscape whose height can be modified at runtime. I created a huge grid with each cell as a section because I didn't want to update whole mesh. When I updated just one section, huge frame drop struck me.

I reimplemented the landscape with PMC and it didn't suffer such problem. But for initial grid creation, RMC was way faster than PMC. Of course I want both operations to be fast, but one beats another for one task and vise versa.

I don't know full details of PMC and RMC implementation, but these are my observations:
- PMC update is not slowed when it has many -like 256 - sections. (collision enabled)
- RMC update is really slowed when it has many sections. (regardless of collision enabled/disabled)
- For CreateSection(), RMC is much faster than PMC.

Besides, from your suggestion it seems it's more appropriate to create a RMC for each cell and let a RMC has just one section. I'll try it and compare to PMC again.
